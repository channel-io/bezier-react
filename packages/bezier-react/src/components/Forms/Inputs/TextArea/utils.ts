import { type SemanticNames } from '~/src/foundation'

interface CalculateBgColorProps {
  focused: boolean
  hasError: boolean
  readOnly: boolean
}

export function getTextAreaBgColorSemanticName({
  focused,
  hasError,
  readOnly,
}: CalculateBgColorProps): SemanticNames {
  if (readOnly) { return 'bg-grey-lighter' }
  if (focused || hasError) { return 'bg-white-normal' }
  return 'bg-grey-lightest'
}
