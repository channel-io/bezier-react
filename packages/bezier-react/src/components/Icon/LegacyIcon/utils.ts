/* Internal Dependencies */
// eslint-disable-next-line no-restricted-imports
import icons, { type IconName } from '../generated'

/**
 * @deprecated Please import and use individual icons.
 * @example
 * import { Icon, AllIcon, type IconProps } from '@channel.io/bezier-react'
 * <Icon source={AllIcon} color="bg-black-dark" />
 * @example <caption>How to validate the bezier icon source</caption>
 * import { isBezierIcon, AllIcon } from '@channel.io/bezier-react'
 * isBezierIcon(AllIcon) // true
 * isBezierIcon(<svg>...</svg>) // false
 * @example <caption>Legacy icon component is still available. but it will be removed in future versions!</caption>
 * import { LegacyIcon, type LegacyIconProps } from '@channel.io/bezier-react'
 * <LegacyIcon name="all" color="bg-black-dark" />
 */
export const isIconName = (name: any): name is IconName => Object.keys(icons).includes(name)
