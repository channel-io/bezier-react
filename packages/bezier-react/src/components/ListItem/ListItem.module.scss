.ListItemLeftIcon {
  transition: color var(--transition-m);
}

.ListItem {
  all: unset;

  cursor: pointer;

  display: flex;
  align-items: center;

  text-decoration: none;

  transition: background-color var(--transition-s);

  &:where(.variant-monochrome) {
    color: var(--txt-black-darkest);

    & :where(.ListItemLeftIcon) {
      color: var(--txt-black-dark);
    }
  }

  &:where(.size-xs) {
    padding: 4px 6px;
    border-radius: var(--radius-6);
  }

  &:where(.size-s) {
    padding: 6px;
    border-radius: var(--radius-6);
  }

  &:where(.size-m) {
    padding: 8px 6px;
    border-radius: var(--radius-8);
  }

  &:where(.size-l) {
    padding: 10px 6px;
    border-radius: var(--radius-12);
  }

  &:where(.disabled) {
    cursor: default;
    opacity: var(--opacity-disabled);
  }

  &:where(:not(.disabled, .active)) {
    &:where(.focused, :focus-visible) {
      background-color: var(--bg-black-lighter);
    }

    &:where(:hover) {
      background-color: var(--bg-black-lighter);
    }
  }

  &:where(.active) {
    color: var(--bgtxt-blue-normal);
    background-color: var(--bgtxt-blue-lightest);

    &:where(.focused, :focus-visible) {
      background-color: var(--bgtxt-blue-lighter);
    }

    & :where(.ListItemLeftIcon) {
      color: var(--bgtxt-blue-normal);
    }

    /* NOTE: When multiple adjacent elements are active, it naturally stitches the border together. */
    /* stylelint-disable selector-max-specificity */
    &:has(+ .ListItem.active) {
      border-bottom-right-radius: 0;
      border-bottom-left-radius: 0;
    }

    & + .ListItem.active {
      border-top-left-radius: 0;
      border-top-right-radius: 0;
    }
    /* stylelint-enable selector-max-specificity */
  }

  &.variant-blue {
    color: var(--bgtxt-blue-normal);

    & :where(.ListItemLeftIcon) {
      color: var(--bgtxt-blue-normal);
    }
  }

  &.variant-red {
    color: var(--bgtxt-red-normal);

    & :where(.ListItemLeftIcon) {
      color: var(--bgtxt-red-normal);
    }
  }

  &.variant-green {
    color: var(--bgtxt-green-normal);

    & :where(.ListItemLeftIcon) {
      color: var(--bgtxt-green-normal);
    }
  }

  &.variant-cobalt {
    color: var(--bgtxt-cobalt-normal);

    & :where(.ListItemLeftIcon) {
      color: var(--bgtxt-cobalt-normal);
    }
  }
}

.ListItemContent {
  display: grid;
  grid-template-columns: fit-content(100%) minmax(0, 1fr);
  width: 100%;
}

.ListItemRightContent {
  display: flex;
  margin-left: 8px;
  white-space: nowrap;
}

.ListItemTitle {
  display: flex;
  grid-column: 2;
  grid-row: 1;
  align-items: center;

  & * {
    transition: color var(--transition-s);
  }
}

.ListItemDescription {
  display: flex;
  grid-column: 2;
  grid-row: 2;
  align-items: center;

  width: 100%;
  margin-top: 2px;
}

.ListItemLeftContent {
  display: flex;
  grid-column: 1;
  grid-row: 1;
  align-items: center;

  margin-right: 8px;
}
