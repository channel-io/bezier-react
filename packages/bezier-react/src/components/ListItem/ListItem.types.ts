import type React from 'react'

import { type BezierIcon } from '@channel.io/bezier-icons'

import type {
  ActivatableProps,
  BezierComponentProps,
  ContentProps,
  DisableProps,
  LinkProps,
  PolymorphicProps,
  SideContentProps,
  SizeProps,
  VariantProps,
} from '~/src/types/props'

export type ListItemSize =
| 'xs'
| 's'
| 'm'
| 'l'

export type ListItemVariant =
| 'blue'
| 'red'
| 'green'
| 'cobalt'
| 'monochrome'

interface ListItemOwnProps {
  focused?: boolean
  description?: React.ReactNode
  descriptionMaxLines?: number
}

export interface ListItemProps extends
  Omit<BezierComponentProps, keyof ContentProps>,
  ContentProps,
  PolymorphicProps,
  SizeProps<ListItemSize>,
  VariantProps<ListItemVariant>,
  SideContentProps<React.ReactNode | BezierIcon, React.ReactNode>,
  LinkProps,
  DisableProps,
  ActivatableProps,
  ListItemOwnProps {}
