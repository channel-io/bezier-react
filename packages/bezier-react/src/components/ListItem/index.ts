export { ListItem } from './ListItem'
export type {
  ListItemProps,
  ListItemSize,
  ListItemVariant,
} from './ListItem.types'

