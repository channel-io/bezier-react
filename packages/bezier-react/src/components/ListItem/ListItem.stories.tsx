/* External dependencies */
import React, { useState, useMemo } from 'react'
import base from 'paths.macro'
import { Meta, Story } from '@storybook/react'

/* Internal dependencies */
import { iconList, getTitle } from 'Utils/storyUtils'
import {
  compact,
} from 'Utils/arrayUtils'
import ListItem from './ListItem'
import ListItemProps, { ListItemSize, ListItemVariant } from './ListItem.types'

export default {
  title: getTitle(base),
  component: ListItem,
} as Meta

const Template: Story<ListItemProps & { width: number }> = ({ width, ...listItemProps }) => (
  <div style={{ width }}>
    <ListItem
      optionKey="menu-item-0"
      {...listItemProps}
    />
  </div>
)

export const Primary = Template.bind({})

Primary.args = {
  width: 388,
  size: ListItemSize.M,
  content: '상담이 열릴 때',
  description: '고객이 첫 메시지를 보내거나, 매니저가 상담을 다시 열거나, 자동으로 리오픈되면 트리거됩니다.',
  rightContent: '',
  leftIcon: 'inbox',
  active: false,
  focused: false,
  disableIconActive: false,
  descriptionMaxLines: 0,
  href: '',
}

Primary.argTypes = {
  width: {
    control: {
      type: 'range',
      min: 50,
      max: 500,
      step: 1,
    },
  },
  disabled: { control: { type: 'boolean' } },
  variant: {
    control: {
      type: 'radio',
      options: [
        ...Object.values(ListItemVariant),
      ],
    },
  },
  active: { control: { type: 'boolean' } },
  leftIcon: { control: { type: 'select', options: [...iconList, undefined] } },
  size: { control: { type: 'select', options: ListItemSize } },
}

interface CompositionProps {
  listRange: number
}

const CompositionTemplate = ({ listRange }: CompositionProps) => {
  const [activeIndex, setActiveIndex] = useState<Set<number>>(() => {
    const randomActiveIndex = Array.from(Array(listRange).keys()).map((index) => (Math.random() < 0.5 ? index : null))
    return new Set(compact<number>([...randomActiveIndex]))
  })

  const isActive = (index: number) => activeIndex.has(index)

  const toggleActive = (index: number) => setActiveIndex((prevSet) => {
    if (prevSet.has(index)) {
      prevSet.delete(index)
      return new Set(Array.from(prevSet))
    }
    return new Set(Array.from(prevSet.add(index)))
  })

  const list = useMemo(() => Array.from(Array(listRange).keys()), [listRange])

  return (
    <div>
      { list.map((index) => (
        <ListItem
          key={index}
          className={isActive(index) ? 'active' : undefined}
          optionKey={`menu-item-${index}`}
          active={isActive(index)}
          onClick={() => toggleActive(index)}
          content={`이것은 ${index}번 아이템입니다.`}
        />
      )) }
    </div>
  )
}

export const Composition: Story<CompositionProps> = CompositionTemplate.bind({})

Composition.args = {
  listRange: 10,
}

Composition.argTypes = {
  listRange: {
    control: {
      type: 'range',
      min: 2,
      max: 20,
    },
  },
}
