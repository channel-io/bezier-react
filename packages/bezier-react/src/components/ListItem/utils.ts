import { css } from '~/src/foundation'

import { ListItemSize } from './ListItem.types'

export function getStyleOfSize(size?: ListItemSize) {
  switch (size) {
    case ListItemSize.S:
      return css`
        padding: 4px 6px;
        ${({ foundation }) => foundation?.rounding.round6};
      `
    case ListItemSize.L:
      return css`
        padding: 8px 6px;
        ${({ foundation }) => foundation?.rounding.round8};
      `
    case ListItemSize.XL:
      return css`
        padding: 10px 6px;
        ${({ foundation }) => foundation?.rounding.round12};
      `
    case ListItemSize.M:
    default:
      return css`
        padding: 6px;
        ${({ foundation }) => foundation?.rounding.round6};
      `
  }
}
