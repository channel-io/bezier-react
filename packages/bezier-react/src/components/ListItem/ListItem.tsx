import React, {
  forwardRef,
  useCallback,
  useMemo,
  useState,
} from 'react'

import classNames from 'classnames'
import { v4 as uuid } from 'uuid'

import useMergeRefs from '~/src/hooks/useMergeRefs'
import { noop } from '~/src/utils/function'
import {
  isEmpty,
  isNil,
  isString,
} from '~/src/utils/type'

import { IconSize } from '~/src/components/Icon'
import { Text } from '~/src/components/Text'

import type ListItemProps from './ListItem.types'
import {
  ListItemSize,
  ListItemVariant,
} from './ListItem.types'
import useAdjacentElementBorderRadius from './useAdjacentElementBorderRadius'

import {
  ContentWrapper,
  Description,
  DescriptionWrapper,
  LeftContentWrapper,
  RightContent,
  StyledIcon,
  Title,
  TitleWrapper,
  Wrapper,
} from './ListItem.styled'

const LINE_BREAK_CHAR = '\n'

export const LIST_ITEM_TEST_ID = 'bezier-react-list-menu-item'

function filterActiveItem(node: HTMLElement) {
  return node.dataset.active === 'true'
}

type ListItemRef = HTMLDivElement & HTMLAnchorElement

function ListItem({
  className,
  contentClassName,
  iconClassName,
  as,
  testId = LIST_ITEM_TEST_ID,
  size = ListItemSize.M,
  descriptionMaxLines,
  content,
  description,
  name,
  leftContent,
  leftIcon,
  variant = ListItemVariant.Monochrome,
  href = '',
  hide = false,
  rightContent = null,
  /* OptionItem Props */
  optionKey,
  /* Activatable Element Props */
  active,
  activeClassName,
  focused = false,
  disabled = false,
  /* HTMLAttribute Props */
  onClick = noop,
  onMouseDown = noop,
  onMouseEnter = noop,
  onMouseLeave = noop,
  ...otherProps
}: ListItemProps,
forwardedRef: React.Ref<ListItemRef>,
) {
  const [listItemElement, setListItemElement] = useState<ListItemRef | null>(null)

  const setListItemRef = useCallback((node: ListItemRef | null) => {
    if (!node) { return }
    setListItemElement(node)
  }, [])

  const mergedRef = useMergeRefs<ListItemRef>(setListItemRef, forwardedRef)

  const isHyperLink = !isEmpty(href)
  const isActive = isHyperLink ? false : active

  useAdjacentElementBorderRadius(listItemElement, filterActiveItem, isActive)

  const mergedClassName = useMemo(() => (
    classNames(className, isActive && activeClassName)
  ), [
    className,
    activeClassName,
    isActive,
  ])

  const handleClick = useCallback((e: React.MouseEvent) => {
    if (!disabled) {
      onClick(e, name)
    }
  }, [
    disabled,
    name,
    onClick,
  ])

  const getNewLineComponent = useCallback((desc: string) => (
    desc.split(LINE_BREAK_CHAR).map((str, index) => {
      if (index === 0) {
        return (
          <Text key={uuid()} typo="12">
            { str }
          </Text>
        )
      }

      return (
        <React.Fragment key={uuid()}>
          <br />
          <Text typo="12">
            { str }
          </Text>
        </React.Fragment>
      )
    })
  ), [])

  const leftComponent = useMemo(() => {
    if (!isNil(leftContent)) {
      return (
        <LeftContentWrapper>
          { leftContent }
        </LeftContentWrapper>
      )
    }

    if (!isNil(leftIcon)) {
      return (
        <LeftContentWrapper>
          <StyledIcon
            className={iconClassName}
            source={leftIcon}
            size={IconSize.S}
            active={isActive}
            variant={variant}
          />
        </LeftContentWrapper>
      )
    }

    return null
  }, [
    isActive,
    iconClassName,
    leftContent,
    leftIcon,
    variant,
  ])

  const titleComponent = useMemo(() => (
    <TitleWrapper className={contentClassName}>
      { isString(content) ? (
        <Title>
          <Text
            typo={size === ListItemSize.XL
              ? '18'
              : '14'}
          >
            { content }
          </Text>
        </Title>
      ) : content }
    </TitleWrapper>
  ), [
    content,
    contentClassName,
    size,
  ])

  const descriptionComponent = useMemo(() => (
    <DescriptionWrapper>
      <Description descriptionMaxLines={descriptionMaxLines}>
        {
          isString(description)
            ? getNewLineComponent(description)
            : description
        }
      </Description>
    </DescriptionWrapper>
  ), [
    description,
    descriptionMaxLines,
    getNewLineComponent,
  ])

  const rightComponent = useMemo(() => (
    <RightContent>
      { rightContent }
    </RightContent>
  ), [rightContent])

  const ContentComponent = useMemo(() => (
    <>
      <ContentWrapper>
        { leftComponent }
        { titleComponent }
        { description && descriptionComponent }
      </ContentWrapper>
      { rightContent && rightComponent }
    </>
  ), [
    leftComponent,
    titleComponent,
    description,
    descriptionComponent,
    rightContent,
    rightComponent,
  ])

  const commonDataAttr = useMemo(() => ({
    'data-active': isActive,
    'data-option-key': optionKey,
    'data-testid': testId,
  }), [
    isActive,
    optionKey,
    testId,
  ])

  const commonListItemProps = useMemo((): ListItemProps & { ref: React.Ref<ListItemRef> } => ({
    ref: mergedRef,
    className: mergedClassName,
    size,
    onClick: handleClick,
    onMouseDown,
    onMouseEnter,
    onMouseLeave,
    active: isActive,
    focused,
    disabled,
    variant,
    ...otherProps,
  }), [
    mergedRef,
    mergedClassName,
    size,
    handleClick,
    onMouseDown,
    onMouseEnter,
    onMouseLeave,
    isActive,
    focused,
    disabled,
    variant,
    otherProps,
  ])

  if (hide) { return null }

  return isHyperLink
    ? (
      <Wrapper
        {...commonDataAttr}
        {...commonListItemProps}
        as={'a' as React.ElementType<any>}
        href={href}
        draggable={false}
        target="_blank"
        rel="noopener noreferrer"
      >
        { ContentComponent }
      </Wrapper>
    )
    : (
      <Wrapper
        {...commonDataAttr}
        {...commonListItemProps}
        as={as}
      >
        { ContentComponent }
      </Wrapper>
    )
}

export default forwardRef(ListItem)
