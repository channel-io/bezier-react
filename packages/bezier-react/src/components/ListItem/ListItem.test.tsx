import React from 'react'

import { render } from '~/src/utils/test'

import { LIST_ITEM_TEST_ID, ListItem } from './ListItem'
import { type ListItemProps } from './ListItem.types'

describe('ListItem', () => {
  let props: ListItemProps

  beforeEach(() => {
    props = {
      content: 'this is content',
      active: false,
    }
  })

  const renderComponent = (optionProps?: Partial<ListItemProps>) =>
    render(
      <ListItem
        {...props}
        {...optionProps}
      />
    )

  it('should have "a tag" related attributes  when "href" prop is string', () => {
    const { getByTestId } = renderComponent({ href: 'https://naver.com' })
    const rendered = getByTestId(LIST_ITEM_TEST_ID)
    expect(rendered).toHaveAttribute('href', 'https://naver.com')
    expect(rendered).toHaveAttribute('rel', 'noopener noreferrer')
    expect(rendered).toHaveAttribute('target', '_blank')
  })
})
