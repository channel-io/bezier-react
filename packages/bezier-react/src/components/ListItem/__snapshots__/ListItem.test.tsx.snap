// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ListItem Snapshot > 1`] = `
.c4 {
  font-size: 1.4rem;
  line-height: 1.8rem;
  margin: 0px 0px 0px 0px;
  font-style: normal;
  font-weight: normal;
  color: inherit;
  -webkit-transition-delay: 0ms;
  transition-delay: 0ms;
  -webkit-transition-timing-function: cubic-bezier(.3,0,0,1);
  transition-timing-function: cubic-bezier(.3,0,0,1);
  -webkit-transition-duration: 150ms;
  transition-duration: 150ms;
  -webkit-transition-property: color;
  transition-property: color;
}

.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  grid-row: 1;
  grid-column: 2;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c2 span {
  -webkit-transition-delay: 0ms;
  transition-delay: 0ms;
  -webkit-transition-timing-function: cubic-bezier(.3,0,0,1);
  transition-timing-function: cubic-bezier(.3,0,0,1);
  -webkit-transition-duration: 150ms;
  transition-duration: 150ms;
  -webkit-transition-property: color;
  transition-property: color;
}

.c3 {
  overflow: hidden;
  text-overflow: ellipsis;
  white-space: nowrap;
}

.c1 {
  display: grid;
  grid-template-columns: fit-content(100%) minmax(0,1fr);
  width: 100%;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  padding: 6px;
  overflow: hidden;
  border-radius: 6px;
  color: #000000D9;
  -webkit-text-decoration: none;
  text-decoration: none;
  cursor: pointer;
  opacity: 1;
  -webkit-transition-delay: 0ms;
  transition-delay: 0ms;
  -webkit-transition-timing-function: cubic-bezier(.3,0,0,1);
  transition-timing-function: cubic-bezier(.3,0,0,1);
  -webkit-transition-duration: 150ms;
  transition-duration: 150ms;
  -webkit-transition-property: background-color;
  transition-property: background-color;
}

.c0:hover {
  background-color: #0000000D;
}

<div
  class="c0"
  data-active="false"
  data-option-key="menu-item"
  data-testid="bezier-react-list-menu-item"
>
  <div
    class="c1"
  >
    <div
      class="c2"
    >
      <div
        class="c3"
      >
        <span
          class="c4"
          data-testid="bezier-react-text"
        >
          this is content
        </span>
      </div>
    </div>
  </div>
</div>
`;
